File: cses/04_graph_algorithms/1192_counting-rooms/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    m = int(data[1])
    rows = data[2:2 + n]

    # Bảng chuyển: '.' (mã 46) -> 1 (sàn), mọi ký tự khác -> 0 (tường).
    table = bytes(1 if c == 46 else 0 for c in range(256))

    # Đệm một viền tường quanh lưới: mảng phẳng (n+2) x (m+2), mặc định toàn 0
    # (tường) để khỏi kiểm tra biên khi mở rộng lân cận. bytes.translate đổi
    # từng dòng '.'/'#' sang 1/0 ở tốc độ C.
    W = m + 2  # chiều rộng sau khi đệm viền
    g = bytearray((n + 2) * W)
    for i in range(n):
        base = (i + 1) * W + 1
        g[base:base + m] = rows[i].translate(table)

    count = 0
    stack = []
    push = stack.append  # bind cục bộ để tăng tốc vòng lặp
    pop = stack.pop
    total = len(g)

    # Duyệt toàn lưới: mỗi ô sàn chưa thăm mở đầu một phòng mới.
    for start in range(total):
        if g[start]:
            count += 1
            g[start] = 0  # đánh dấu đã thăm bằng cách đặt lại thành tường
            push(start)
            # Flood fill (DFS lặp) quét trọn thành phần liên thông hiện tại,
            # xét lần lượt 4 lân cận phải/trái/trên/dưới.
            while stack:
                cur = pop()
                r = cur + 1
                if g[r]:
                    g[r] = 0
                    push(r)
                l = cur - 1
                if g[l]:
                    g[l] = 0
                    push(l)
                u = cur - W
                if g[u]:
                    g[u] = 0
                    push(u)
                d = cur + W
                if g[d]:
                    g[d] = 0
                    push(d)

    sys.stdout.write(str(count) + "\n")


main()

File: cses/04_graph_algorithms/1192_counting-rooms/solution.cpp
#include <iostream>
#include <string>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, m;
    std::cin >> n >> m;

    // Đệm một viền tường quanh lưới: mảng phẳng (n+2) x (m+2),
    // mặc định toàn 0 (tường) để khỏi kiểm tra biên khi mở rộng lân cận.
    const int width = m + 2;
    std::vector<char> floor((n + 2) * width, false);
    for (int row = 1; row <= n; ++row) {
        std::string line;
        std::cin >> line;
        for (int column = 1; column <= m; ++column) {
            // '.' -> sàn (true = 1), '#' -> tường (false = 0)
            floor[row * width + column] = (line[column - 1] == '.');
        }
    }

    int rooms = 0;
    std::vector<int> stack;
    stack.reserve(n * m);

    // Duyệt toàn lưới: mỗi ô sàn chưa thăm mở đầu một phòng mới.
    for (int start = 0; start < static_cast<int>(floor.size()); ++start) {
        if (!floor[start]) {
            continue;
        }

        ++rooms;
        floor[start] = false;  // đánh dấu đã thăm bằng cách đặt lại thành tường
        stack.push_back(start);

        // Flood fill (DFS lặp) quét trọn thành phần liên thông hiện tại.
        while (!stack.empty()) {
            const int cell = stack.back();
            stack.pop_back();

            // 4 lân cận: trái, phải, trên, dưới (nhờ viền, không cần kiểm biên).
            const int neighbors[4] = {cell - 1, cell + 1, cell - width,
                                      cell + width};
            for (int neighbor : neighbors) {
                if (floor[neighbor]) {
                    floor[neighbor] = false;
                    stack.push_back(neighbor);
                }
            }
        }
    }

    std::cout << rooms << '\n';
    return 0;
}
